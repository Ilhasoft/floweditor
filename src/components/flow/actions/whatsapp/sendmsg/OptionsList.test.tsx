import { mock } from 'testUtils';
import OptionsList from './OptionsList';
import * as utils from 'utils';
import * as React from 'react';
import { act, fireUnnnicInputChangeText, render, wait } from 'test/utils';

mock(utils, 'createUUID', utils.seededUUIDs());

function getProps() {
  return {
    listTitle: {
      value: 'list title'
    },
    listFooter: {
      value: 'list footer'
    },
    options: {
      value: [
        { uuid: '1', title: 'title 1', description: 'description 1' },
        { uuid: '2', title: 'title 2', description: 'description 2' }
      ],
      validationFailures: [] as any[]
    },
    onOptionsUpdated: jest.fn(),
    onListTitleUpdated: jest.fn(),
    onListFooterUpdated: jest.fn(),
    onOptionRemoval: jest.fn()
  };
}

describe(OptionsList.name, () => {
  it('should render', async () => {
    const props = getProps();
    const { baseElement } = render(<OptionsList {...props} />);

    await wait();

    expect(baseElement).toMatchSnapshot();
  });

  it('should update list title', async () => {
    const props = getProps();
    const { getByTestId } = render(<OptionsList {...props} />);

    await wait();

    const listTitleInput = getByTestId('List title (optional)');
    await act(async () => {
      fireUnnnicInputChangeText(listTitleInput, 'new list title');
    });

    expect(props.onListTitleUpdated).toHaveBeenCalledWith('new list title');
  });

  it('should update list footer', async () => {
    const props = getProps();
    const { getByTestId } = render(<OptionsList {...props} />);

    await wait();

    const listFooterInput = getByTestId('Footer (optional)');
    await act(async () => {
      fireUnnnicInputChangeText(listFooterInput, 'new list footer');
    });

    expect(props.onListFooterUpdated).toHaveBeenCalledWith('new list footer');
  });

  it('should update list item title', async () => {
    const props = getProps();
    const { getByTestId, getAllByTestId } = render(<OptionsList {...props} />);

    await wait();

    // expand collapse
    const expandCollapseButton = getByTestId('List options');
    await act(async () => {
      expandCollapseButton.click();
    });

    const listItemTitleInput = getAllByTestId('Title')[1];
    await act(async () => {
      fireUnnnicInputChangeText(listItemTitleInput, 'new title 2');
    });

    expect(props.onOptionsUpdated).toHaveBeenCalledWith([
      { uuid: '1', title: 'title 1', description: 'description 1' },
      { uuid: '2', title: 'new title 2', description: 'description 2' }
    ]);
  });

  it('should update list item description', async () => {
    const props = getProps();
    const { getByTestId, getAllByTestId } = render(<OptionsList {...props} />);

    await wait();

    // expand collapse
    const expandCollapseButton = getByTestId('List options');
    await act(async () => {
      expandCollapseButton.click();
    });

    const listItemDescriptionInput = getAllByTestId('Description (Optional)')[1];
    await act(async () => {
      fireUnnnicInputChangeText(listItemDescriptionInput, 'new description 2');
    });

    expect(props.onOptionsUpdated).toHaveBeenCalledWith([
      { uuid: '1', title: 'title 1', description: 'description 1' },
      { uuid: '2', title: 'title 2', description: 'new description 2' }
    ]);
  });

  it('should remove list item', async () => {
    const props = getProps();
    const { getByTestId, getAllByTestId } = render(<OptionsList {...props} />);

    await wait();

    // expand collapse
    const expandCollapseButton = getByTestId('List options');
    await act(async () => {
      expandCollapseButton.click();
    });

    const listItemRemoveButton = getAllByTestId('Remove')[1];
    await act(async () => {
      listItemRemoveButton.click();
    });

    expect(props.onOptionRemoval).toHaveBeenCalledWith({
      uuid: '2',
      title: 'title 2',
      description: 'description 2'
    });
  });

  it('should not remove last list item if it is an empty one', async () => {
    const props = getProps();
    props.options.value = [{ uuid: '1', title: '', description: '' }];
    const { getByTestId, getAllByTestId } = render(<OptionsList {...props} />);

    await wait();

    // expand collapse
    const expandCollapseButton = getByTestId('List options');
    await act(async () => {
      expandCollapseButton.click();
    });

    const listItemRemoveButton = getAllByTestId('Remove')[0];
    await act(async () => {
      listItemRemoveButton.click();
    });

    expect(props.onOptionRemoval).not.toHaveBeenCalled();
  });

  it('should show error message', () => {
    const props = getProps();
    props.options.validationFailures = [{ message: 'error message text' }];
    const { getByTestId } = render(<OptionsList {...props} />);

    const errorMessage = getByTestId('Error message');

    expect(errorMessage).toBeInTheDocument();
  });
});
