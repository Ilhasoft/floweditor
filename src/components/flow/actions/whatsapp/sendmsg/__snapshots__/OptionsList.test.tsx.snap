// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsList should render 1`] = `
<body>
  <div>
    <div
      class="list_wrapper"
    >
      <div
        class="list_inputs"
      >
        <div>
          <span
            class="label"
          >
            List title (optional)
          </span>
          <div
            class="completion_wrapper"
          >
            <div
              data-testid="List title (optional)"
            >
              <div
                data-use-vue-component-wrap=""
                style="all: unset;"
              >
                <div
                  class="unnnic-form md"
                  data-passed-props="[object Object]"
                  data-v-48c0c241=""
                >
                  <!---->
                  <div
                    class="unnnic-form-input text-input size--md"
                    data-v-400546b3=""
                    data-v-48c0c241=""
                    mask=""
                    v-bind="[object Object]"
                  >
                    <input
                      class="input-itself input size-md normal"
                      data-passed-props="[object Object]"
                      data-v-400546b3=""
                      data-v-97b44e8c=""
                      placeholder="Title"
                      size="10"
                      type="text"
                      v-bind="[object Object]"
                    />
                    <!---->
                    <!---->
                  </div>
                  <!---->
                </div>
              </div>
            </div>
            <div
              class="options_wrapper"
              data-testid="temba_completion_options_List title (optional)"
              style="margin-top: 0px; width: -24px; display: none;"
            />
          </div>
        </div>
        <div>
          <span
            class="label"
          >
            Footer (optional)
          </span>
          <div
            class="completion_wrapper"
          >
            <div
              data-testid="Footer (optional)"
            >
              <div
                data-use-vue-component-wrap=""
                style="all: unset;"
              >
                <div
                  class="unnnic-form md"
                  data-passed-props="[object Object]"
                  data-v-48c0c241=""
                >
                  <!---->
                  <div
                    class="unnnic-form-input text-input size--md"
                    data-v-400546b3=""
                    data-v-48c0c241=""
                    mask=""
                    v-bind="[object Object]"
                  >
                    <input
                      class="input-itself input size-md normal"
                      data-passed-props="[object Object]"
                      data-v-400546b3=""
                      data-v-97b44e8c=""
                      placeholder="Footer text"
                      size="11"
                      type="text"
                      v-bind="[object Object]"
                    />
                    <!---->
                    <!---->
                  </div>
                  <!---->
                </div>
              </div>
            </div>
            <div
              class="options_wrapper"
              data-testid="temba_completion_options_Footer (optional)"
              style="margin-top: 0px; width: -24px; display: none;"
            />
          </div>
        </div>
      </div>
      <div>
        <div
          class="collapse "
        >
          <div
            class="collapse_header"
            data-testid="List options"
          >
            <span>
              List options
            </span>
            <div
              data-use-vue-component-wrap=""
              style="all: unset;"
            >
              <span
                class="material-symbols-rounded unnnic-icon-scheme--neutral-darkest unnnic-icon-size--md"
                data-passed-props="[object Object]"
                data-v-ab84ef7c=""
              >
                 expand_more 
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
